proposals/0000-package-internals.rst

```diff
@@ -0,0 +1,148 @@
+Accessing package internal modules
+==================================
+
+.. author:: Richard Eisenberg and Simon Peyton Jones
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/528>`_.
+.. sectnum::
+.. contents::
+
+Package authors include an ``exposed-modules`` stanza in ``.cabal`` files to
+list the modules available for import outside the package. Other modules are
+not available for import. Yet it is sometimes useful to access even unexposed
+("internal") modules. This proposal offers a way to do so.
+
+Note that internal modules of a package are *not* a part of that package's
+published API and thus changes therein would *not* require a major version
+bump. These modules do remain internal!
+
+Motivation
+----------
+
+1. Suppose package ``pack`` has several internal modules, whose exports are
+   not intended for the public API of ``pack``. Yet the author of ``pack``
+   still wants to test functions exported by these internal modules.
+
+   The author currently has two options, both bad:
+
+   1. Expose the internal modules, perhaps naming them with ``Internal`` or
+      similar.
+
+   2. Structure the project to recompile the entire package in order to run
+      the testsuite, instead of using the already-compiled package.
+
+   This proposal allows the modules to remain internal while still being
+   available for testing without recompilation.
+
+2. Suppose package ``pack`` exports a very general interface to a particular
+   algorithm. Because it is so general, ``pack`` offers no way of, say, using
+   real-world data to seed the algorithm. Instead, downstream packages such as
+   ``pack-json`` and ``pack-csv`` provide such access.
+
+   If an internal module in ``pack`` needs testing, then there is no choice but
+   to expose the internal module, so it can be tested in a downstream package.
+
+3. Suppose two packages ``pack1`` and ``pack2`` are developed together, with
+   ``pack2`` depending on ``pack1``. Because the two packages form logically one
+   whole (but are separated because, say, ``pack1`` has many fewer dependencies
+   than ``pack2`` or other reasons), ``pack2`` needs to import internal modules
+   from ``pack1``. Yet these internal modules are not intended to form part of
+   any public API.
```

Comment:
The blessed way to solve this is a package, containing multiple (public or internal) libraries. This has been implemented in Cabal 3.8, pending Hackage support.